core/src/main/scala/org/apache/spark/scheduler/Pool.scala

```diff
@@ -119,4 +119,66 @@ private[spark] class Pool(
       parent.decreaseRunningTasks(taskNum)
     }
   }
+
+  // Update the number of slots considered available for each TaskSetManager whose ancestor
+  // in the tree is this pool
+  // For FAIR scheduling, slots are distributed among pools based on weights and minshare.
+  //   If a pool requires fewer slots than are available to it, the leftover slots are redistributed
+  //   to the remaining pools using the remaining pools' weights.
+  // For FIFO scheduling, the schedulable queue is iterated over in FIFO order,
+  //   giving each schedulable the remaining slots,
+  //   up to the number of remaining tasks for that schedulable.
+  override def updateAvailableSlots(numSlots: Float): Unit = {
+    schedulingMode match {
+      case SchedulingMode.FAIR =>
+        val queueCopy = new ConcurrentLinkedQueue[Schedulable](schedulableQueue)
+        var shouldRedistribute = false
+        var totalWeights = schedulableQueue.asScala.map(_.weight).sum
+        var totalSlots = numSlots
+        do {
+          shouldRedistribute = false
+          var nextWeights = totalWeights
+          var nextSlots = totalSlots
+          queueCopy.forEach(schedulable => {
+            val numTasksRemaining = schedulable.getSortedTaskSetQueue
+              .map(tsm => tsm.tasks.length - tsm.tasksSuccessful).sum
+            val allocatedSlots = Math.max(
+              totalSlots * schedulable.weight / totalWeights,
+              schedulable.minShare)
+            if (numTasksRemaining < allocatedSlots) {
+              schedulable.updateAvailableSlots(numTasksRemaining)
+              nextWeights -= schedulable.weight
+              nextSlots -= numTasksRemaining
+              shouldRedistribute = true
```

Comment:
this `shouldRedistribute` seems unnecessary, shall we just write `while (totalSlots > 0)` and keep decreasing `totalSlots` in the loop body?

Reply:
I couldn't yet think of a way to get rid of the `shouldRedistribute `
However, I've added the check on `totalSlots` as a shortcut